packages/itwinui-react/src/core/ComboBox/ComboBox.tsx

```diff
@@ -438,16 +472,16 @@ export const ComboBox = React.forwardRef(
         isMenuItemSelected,
         multiple,
         onChangeHandler,
-        selected,
         optionsRef,
         hide,
+        setSelectedIndexes,
       ],
     );
 
     const getMenuItem = React.useCallback(
       (option: SelectOption<T>, filteredIndex?: number) => {
         const optionId = getOptionId(option, id);
-        const { __originalIndex } = optionsExtraInfoRef.current[optionId];
+        const { __originalIndex } = optionsExtraInfo[optionId];
```

Comment:
unit tests failing here

Reply:
Related: https://github.com/iTwin/iTwinUI/discussions/2070

Reply:
In https://github.com/iTwin/iTwinUI/pull/2048#discussion_r1630182367, you said:

> I believe that updating `optionsExtraInfo` directly inside the render function or a `useMemo` is a problem as that immediately updates `optionsExtraInfo` based on the _new_ options even though the `filteredOptions` being used in the return part of the render function is still referring to the _old_ options. This is why when `getMenuItem` tries to search for _old_ options in `optionsExtraInfo` during the render, no match is found.

…which makes me wonder: Shouldn't we fix `filteredOptions` to update eagerly instead of adding a workaround for it (i.e. more stale state)?

I can understand that fixing `filteredOptions` could be considered out-of-scope for this PR, but then it begs the question: why did we even introduce `optionsExtraInfo` in the first place? Was there a problem with keeping it as `optionsExtraInfoRef`?

Reply:
> Yes, with a ref, it updates it immediately in the render function itself instead of after the first render.

This does not explain why it's a problem now and wasn't a problem before.
 
> > Shouldn't we fix `filteredOptions` to update eagerly instead of adding a workaround for it (i.e. more stale state)?
> 
> I think I had tried doing this by calling `setFilteredOptions` directly inside the render function instead of in a `useEffect`/`useLayoutEffect`. 

No, that's still indirect/stale state. To do it eagerly, you can use `useMemo` (same as my `optionsExtraInfo` suggestion).

```jsx
const filteredOptions = React.useMemo(() => {
  return filterFunction(options, value);
}, [options, inputValue, filterFunction]);
```

Reply:
This was fixed during pairing session where we moved both `optionsExtraInfo` and `filteredOptions` into `useMemo`s that re-compute whenever `options` change [`b768fc5`](https://github.com/iTwin/iTwinUI/pull/2048/commits/b768fc5ca8b3d42a49986cc9a93915e7562d7936)